### ISSUE TITLE
Lint for named, but unused, futures

### ISSUE THREAD

**Issue by goffrie** [rust-lang/rust-clippy#8126]
### What it does

In some situations, such as [async-aware Mutexes](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html), it is useful to call an async function and bind its result to a variable so that it drops at the end of the scope - e.g. `let _guard = mutex.lock().await`. However, it's probably a mistake to do so _without_ awaiting (or `block_on`, or otherwise consuming) the Future.

This is a bit rare as you'd usually notice when making use of the returned guard, but not always. This came up today and was a real head-scratcher.

### Category

correctness

### Drawbacks

This could cause a false positive when calling a (non-`async`) function that returns a type that implements `Future` but also has its own nontrivial `Drop` implementation, that the caller intentionally wants to defer. I'm not aware of any use case where that would apply. In particular, this never applies to an `async fn` as dropping an `async` block does nothing except drop its captured arguments.

### Example

```rust
pub async fn doit(x: std::sync::Arc<tokio::sync::Mutex<()>>) {
    let _foo = x.lock();
    println!("i'm exclusive");
}
```

Should be instead:

```rust
pub async fn doit(x: std::sync::Arc<tokio::sync::Mutex<()>>) {
    let _foo = x.lock().await; // `.await` here
    println!("i'm exclusive");
}
```

**Comment by camsteffen** [comment]
So basically detect a `Future` assigned to a variable but then never used? Name idea: `unused_future_binding`.
